### DISCUSSION TITLE
Construct models without using the alias attribute name

### DISCUSSION THREAD

**Discussion by daemonkeeper** [pydantic/pydantic#10775] [Pydantic V2]
Hi,
I'm a bit lost. I thought I understood Pydantic's [field aliases](https://docs.pydantic.dev/latest/concepts/fields/#field-aliases) as an additional way to reference attributes when constructing a model, without shadowing the original name?

For instance, take this example 

```python
from pydantic import BaseModel, Field, ConfigDict


class User(BaseModel):
    full_name: str = Field(alias='full-name')

user = User(full_name="johndoe")
print(user)
print(user.model_dump(by_alias=True))
```

I would expect this to work as is. However, I get a validation error:

```
1 validation error for User
full-name
  Field required [type=missing, input_value={'full_name': 'johndoe'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.9/v/missing
```

That said, this works:

```python
u = {"full-name": "johndoe"}
user = User(**u)
print(user)
print(user.model_dump(by_alias=True))
```

gives

```
full_name='johndoe'
{'full-name': 'johndoe'}
```

Even funnier if I set 

```python
class User(BaseModel):
    model_config = ConfigDict(extra="forbid")
    full_name: str = Field(alias='full-name')
```

This yields

```
pydantic_core._pydantic_core.ValidationError: 2 validation errors for User
full-name
  Field required [type=missing, input_value={'full_name': 'johndoe'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.9/v/missing
full_name
  Extra inputs are not permitted [type=extra_forbidden, input_value='johndoe', input_type=str]
    For further information visit https://errors.pydantic.dev/2.9/v/extra_forbidden
```

Do I simply misunderstand what and how alias is supposed to work, or is this maybe a bug?

For context, what I want to achieve is that serialized data (e.g. in the YAML dump) uses hyphens for keys, but internally, in the model, the objects all map to field names with underscores.  From https://github.com/pydantic/pydantic/issues/2266 I get the impression that the above should work? 

I've tested the code above with various versions in the 2.x branch of Pydantic. The actual example is tested with 2.9.2

```
$ pip3 freeze | grep pydantic
pydantic==2.9.2
pydantic_core==2.23.4
```

**Comment by Viicos** [comment]
You want https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name.

**Reply by daemonkeeper** [reply]
Thanks, that solves the issue. I wonder how I missed that. 

Besides, I'd argue that should be the default.

**Reply by glebignatieff** [reply]
Be careful, because with `populate_by_name` your model now accepts both `full_name` and `full-name` fields. Sometimes, it's not what you want. You might want class properties to adhere Python's snake case notation and then validate camelCase APIs.

**Reply by Viicos** [reply]
Since 2.11, you can set `validate_by_name` to `True` (replacing `populate_by_name`), and `validate_by_alias=False`.

**Reply by isomorpheme** [reply]
I have the same question, and I don't see how `validate_by_name` solves the issue entirely: I have a field `foo_bar` in Python and I want to use that name when calling the model's constructor, but when validating JSON or a dict I only want to allow camelcase `fooBar`. If I understand correctly, `validate_by_name=True` is not fine-grained enough, allowing `fooBar` *and* `foo_bar` when validating.

**Reply by Viicos** [reply]
Yes, this is a current limitation. See the two closely related threads: https://github.com/pydantic/pydantic/issues/12056 / https://github.com/pydantic/pydantic/issues/11857.
